### PR TITLE
add counting ability on client

### PR DESCRIPTION
**Pull request by e7mac**

### CODE REVIEW

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 1]
```diff
@@ -814,6 +814,15 @@ protected Void doInBackground(Void... params) {
...
 
+            JSONArray enabledCounters = response.optJSONArray(
+                    Constants.Keys.ENABLED_COUNTERS);
+            if (enabledCounters != null) {
+              for (int i=0; i < enabledCounters.length(); i++) {
+                String name = enabledCounters.getString(i);
+                Log.enableCounter(name);
+              }
+            }
+
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: This `handleStartResponse` is already wayyyy too long. Can you take this bit of code out to something like `checkCounters(response)` or something.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumUIEditorWrapper.java [hunk 2]
```diff
@@ -63,6 +63,7 @@ protected LeanplumUIEditorWrapper() {
           interfaceEditorSingleton = (LeanplumUIEditor) method.invoke(null);
           if (interfaceEditorSingleton != null) {
             interfaceEditorSingleton.allowInterfaceEditing(Constants.isDevelopmentModeEnabled);
+            Log.count("sdkCounterUiEditor");
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/LeanplumUIEditorWrapper.java]: `sdkCountUiEditor` (no -er) ? It's shorter and in the design doc and just as accurate, IMHO

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 3]
```diff
@@ -814,6 +814,15 @@ protected Void doInBackground(Void... params) {
               Constants.loggingEnabled = true;
             }
 
+            JSONArray enabledCounters = response.optJSONArray(
+                    Constants.Keys.ENABLED_COUNTERS);
+            if (enabledCounters != null) {
+              for (int i=0; i < enabledCounters.length(); i++) {
+                String name = enabledCounters.getString(i);
+                Log.enableCounter(name);
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: Why not
```Java
Set counterSet = new HashSet<>(Arrays.asList(enabledCounters));
Log.setEnabledCounters(counterSet);
```
Or make `setEnabledCounters` something that takes an array? Simplifies some logic.

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 4]
```diff
@@ -2142,4 +2146,14 @@ private static void parseVariantDebugInfo(JSONObject response) {
...
+  private static void parseSdkCounters(JSONObject response) {
+
+    JSONArray enabledCounters = response.optJSONArray(
+            Constants.Keys.ENABLED_COUNTERS);
+    if (enabledCounters != null) {
+      Set counterSet = new HashSet<>(Arrays.asList(enabledCounters));
+      Log.setEnabledCounters(counterSet);
+    }
+  }
+}
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: nit: empty line at end of file is more tidy. You can tell by Github's bias in design showing that it's missing with a red circle and cross.

**Comment by e7mac** [comment]
test this please

**Comment by e7mac** [comment]
test this please

File: AndroidSDKCore/src/main/java/com/leanplum/internal/Log.java [hunk 5]
```diff
@@ -141,6 +144,7 @@ public static void log(LeanplumLogType type, String message) {
     }
   }
 
+
```
Comment (e7mac) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/Log.java]: this is probably stray whitespaces. Let's clean these up.

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 6]
```diff
@@ -2158,4 +2163,14 @@ private static void parseVariantDebugInfo(JSONObject response) {
   public static void clearUserContent() {
     VarCache.clearUserContent();
   }
+
+  private static void parseSdkCounters(JSONObject response) {
+
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: extra line

File: AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java [hunk 7]
```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2016, Leanplum, Inc. All rights reserved.
```
Comment (MaitreyaBuddha) [inline AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java]: 2018?

File: AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+    countAggregator.incrementCount(testString);
+    HashMap<String, Integer> counts = countAggregator.getCounts();
+    int count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(0, count);
+
+    countAggregator.incrementCount(testString);
+    counts = countAggregator.getCounts();
+    count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(0, count);
+
```
Comment (MaitreyaBuddha) [inline AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java]: extra line

File: AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+    int count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(2, count);
+
+    countAggregator.incrementCount(testString, 15);
+    counts = countAggregator.getCounts();
+    count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(17, count);
+
+  }
+}
```
Comment (MaitreyaBuddha) [inline AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java]: Our style is to always have a newline at the end of files.

File: AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+
+    countAggregator.incrementCount(testString, 2);
+    HashMap<String, Integer> counts = countAggregator.getCounts();
+    int count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(0, count);
+
+    countAggregator.incrementCount(testString, 15);
+    counts = countAggregator.getCounts();
+    count = counts.containsKey(testString) ? counts.get(testString) : 0;
+    assertEquals(0, count);
```
Comment (MaitreyaBuddha) [inline AndroidSDKTests/src/test/java/com/leanplum/internal/CountAggregatorTest.java]: I don't understand why this is 0, or what we are testing here. Tests should be extremely deterministic. We should know exactly if `counts.containsKey(testString)` here, we should not be testing if it exists. If we know it doesn't exist, we are just testing if `count = 0` works. If it does exist, then shouldn't this be `assertEquals(15, count);`?

File: AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+    }
+
+    private HashMap<String, Integer> getAndClearCounts() {
+        HashMap<String, Integer> previousCounts = new HashMap<>();
+        previousCounts.putAll(counts);
+        counts.clear();
+        return previousCounts;
+    }
+
+    public void sendAllCounts() {
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java]: Can we test this? If not, why? Ideally we could mock Request and test that it receives the correct API calls.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/Log.java [hunk 12]
```diff
@@ -63,7 +63,8 @@
      * Used only for Leanplum SDK debugging. Not visible to customers. Not sent to us when remote
      * logging is enabled.
      */
-    DEBUG
+    DEBUG,
+
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/Log.java]: We don't need to commit non-functional changes. Please remove anything that isn't in scope of the feature.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/Request.java [hunk 13]
```diff
@@ -581,6 +581,7 @@ private RequestsWithEncoding getRequestsWithEncodedString() {
   }
 
   private void sendRequests() {
+    CountAggregator.INSTANCE.sendAllCounts();
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/Request.java]: Let's discuss throttling this to be less frequent. If we send half as often it means we can increase our sample size 2x.

File: AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java [hunk 14]
```diff
@@ -0,0 +1,62 @@
+package com.leanplum.internal;
+
+import java.util.HashSet;
+import java.util.HashMap;
+import java.util.Map;
+
+public class CountAggregator {
+
+    public static final CountAggregator INSTANCE = new CountAggregator();
```
Comment (MaitreyaBuddha) [inline AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java]: I'm not sure I like this paradigm. It's minimal and flexible _now_ but is a slippery slope. Maybe we could have a factory?

File: AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java (hunk 11 above)
Comment (gg4race) [reply AndroidSDKCore/src/main/java/com/leanplum/internal/CountAggregator.java]: we will add a test for this in the future

**Pull request merged**